src/printer/tvmscript_unified_printer.cc

```diff
@@ -0,0 +1,738 @@
+  }
+
+  Doc operator()(const ObjectRef& n, TVMScriptUnifiedPrinter& p) const {
+    ICHECK(can_dispatch(n)) << "DocProducerRegistry calls un-registered function on type "
+                            << n->GetTypeKey();
+    return producers_[n->type_index()](n, p);
+  }
+
+  template <typename NodeRef, typename DocType>
+  DocProducerRegistry& register_producer(ProducerType<NodeRef, DocType> producer) {
+    using NodeType = typename NodeRef::ContainerType;
+    uint32_t tindex = NodeType::RuntimeTypeIndex();
+    if (producers_.size() <= tindex) {
+      producers_.resize(tindex + 1, nullptr);
+    }
+    ICHECK(producers_[tindex] == nullptr)
+        << "Dispatch for " << NodeType::_type_key << " is already set";
+
+    producers_[tindex] = [producer = std::move(producer)](const ObjectRef& ref,
+                                                          TVMScriptUnifiedPrinter& p) {
+      Doc doc = producer(Downcast<NodeRef>(ref), p);
+      return doc;
+    };
+    return *this;
+  }
+};
+
+class TVMScriptUnifiedPrinter {
+ public:
+  explicit TVMScriptUnifiedPrinter(std::unique_ptr<DocPrinter> element_printer)
+      : doc_printer_(std::move(element_printer)){};
+
+  static DocProducerRegistry& registry();
+
+  String PrintNode(const ObjectRef& ref);
+
+  template <typename T, typename = std::enable_if_t<std::is_base_of<Doc, T>::value>>
+  T ToDoc(const ObjectRef& ref);
+
+  template <typename DocType, typename NodeType,
+            typename = std::enable_if_t<std::is_base_of<Doc, DocType>::value>,
+            typename = std::enable_if_t<std::is_base_of<ObjectRef, NodeType>::value>>
+  Array<DocType> ToDocArray(const Array<NodeType>& refs);
+
+  ExprDoc ToExprDoc(const ObjectRef& ref) { return ToDoc<ExprDoc>(ref); }
+
+  template <typename NodeType,
+            typename = std::enable_if_t<std::is_base_of<ObjectRef, NodeType>::value>>
+  Array<ExprDoc> ToExprDocArray(const Array<NodeType>& refs) {
+    return ToDocArray<ExprDoc>(refs);
+  }
+
+  Doc PrintExtraVarDeclaration();
+
+  TypeDoc GetBufferTypeDoc(const Buffer& buf);
+  TypeDoc GetVarTypeDoc(const Var& var);
+
+  void OnVarUsed(const Var& var) {
+    auto& name = var->name_hint;
+    if (!context_manager->GetVar(name) && !HasFreeVar(name, var)) {
+      AssignDoc declaration;
+      declaration->target = IdentifierDoc(name);
+      declaration->type = GetVarTypeDoc(var);
+      prelude_.push_back(std::move(declaration));
+      free_vars_.Set(name, var);
+    }
+  }
+
+  void OnBufferUsed(const Buffer& buffer) {
+    auto& name = buffer->name;
+    if (!context_manager->GetVar(name)) {
+      AssignDoc declaration;
+      declaration->target = IdentifierDoc(name);
+      declaration->type = GetBufferTypeDoc(buffer);
+      prelude_.push_back(std::move(declaration));
+      free_vars_.Set(name, buffer);
+    }
+  }
+
+  PrinterContextManager context_manager;
+
+ protected:
+  std::unique_ptr<DocPrinter> doc_printer_;
+  Array<StmtDoc> prelude_;
```

Comment:
tentative syntax

option 1.

```
@T.fragment
def tir_fragment():
  a: T.var("int32")
  def body():
    for ...:
      ...
  return body
```

option 2.

```
a: T.var("int32")
for ...:
  ...
```

Reply:
open question: T.int32 vs T.var(...). The left one is more pythonic, while the right one is more TVM-ish